feat: pc_sign and code_challenge implementation

Gustash opened the pull request:
## Why

The current codebase hardcodes a pc_sign, and uses the `token` authentication flow instead of `code`.

## Changes

- [x] Implemented pc_sign logic, fully compliant with EA Desktop's implementation;
- [x] Generating login URL using `code` instead of `token` strategy;
- [x] Handling `code` strategy.

Gustash committed:
feat: implemented pc_sign

Gustash committed:
fix: encoding of pc_sign

Gustash committed:
fix: remove unused import

battledash commented on maxima-lib/src/core/auth/context.rs:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    pub fn get_url(&self) -> String {
+        let basic_url = format!("{}?code_challenge_method=S256&client_id=JUNO_PC_CLIENT&response_type=code%20id_token&redirect_uri=qrc:///html/login_successful.html&display=junoClient/login&locale=en_US&sbiod_enabled=true", API_NUCLEUS_AUTH);
```
Ideally this should be built dynamically, like at `src/core/auth/mod.rs:18`

battledash replied:
Also, sbiod_enabled should be False. It causes external auths (xbox, etc) to open a new browser window, which is only needed if we're in a webview

battledash commented on maxima-lib/src/core/auth/login.rs:
```diff
@@ -99,5 +104,6 @@ pub async fn manual_login(persona: &str, password: &str) -> Result<String> {
     let token = error.auth_token.unwrap().value;
-    execute_auth_exchange(&token, "JUNO_PC_CLIENT", "token").await
+    let pc_sign = AuthContext::new()?.generate_pc_sign();
+    execute_auth_exchange(&token, "JUNO_PC_CLIENT", &pc_sign, "token").await
```
This should be `code`, not `token`. the final parameter is `response_type`, and since we already have a full token here we need to exchange it back to a code which we can then get an access token and refresh token from.

imLinguin replied:
We don't have code_verifier to complete the token flow. As far as I know `connect/token` needs it

battledash commented on maxima-lib/src/core/auth/mod.rs:
```diff
@@ -10,13 +12,14 @@ use super::endpoints::{API_NUCLEUS_AUTH, API_NUCLEUS_TOKEN};
     let query = vec![
         ("client_id", client_id),
         ("response_type", response_type),
         // Need to figure out how this is generated, see login.rs for more info
```
Remove the comment here since it's irrelevant now

battledash commented on maxima-lib/src/lsx/request/auth.rs:
```diff
@@ -15,11 +15,12 @@ pub async fn handle_auth_code_request(
     info!("Retrieving authorization code for '{}'", client_id);
 
-    let auth_code = execute_auth_exchange(&access_token, &client_id, "code")
+    let auth_code = execute_auth_exchange(&access_token, &client_id, &pc_sign, "code")
```
Would it not be easier to put the `generate_pc_sign` call inside `execute_auth_exchange`? We need to call `execute_auth_exchange` in a variety of places and it's not like the pc_sign will need to change between them

battledash commented on maxima-lib/src/core/auth/context.rs:
```diff
@@ -0,0 +1,74 @@
+        self.code_verifier.as_str()
+    }
+
+    pub fn get_url(&self) -> String {
```
The general convention for getters/setters in rust is to exclude `get_`, and have only `set_`. So this should be `code_verifier()` and `url()`.

battledash commented on maxima-lib/src/core/auth/context.rs:
```diff
@@ -0,0 +1,74 @@
+                ("code_challenge", self.code_challenge.as_str()),
+                ("pc_sign", signature.as_str()),
+                ("nonce", random::<i32>().to_string().as_str()),
+            ],
```
Borrow instead of calling `as_str` here. `&basic_url,` `&self.code_challenge,` etc.

battledash commented on maxima-lib/src/core/auth/context.rs:
```diff
@@ -0,0 +1,74 @@
+
+    /// Returns String representation of code_verifier
+    pub fn get_code_verifier(&self) -> &str {
+        self.code_verifier.as_str()
```
Borrow here, `&self.code_verifier`

imLinguin committed:
feat: access token exchange

applied review comments, changed begin_oauth_flow to return AuthContext which then can be used to get tokens

imLinguin committed:
fix: auth flow in ui

imLinguin committed:
fix: native build

battledash commented on maxima-lib/src/core/auth/mod.rs:
```diff
@@ -77,53 +80,37 @@ pub struct TokenResponse {
+        ("client_id", "JUNO_PC_CLIENT"),
+        (
+            "client_secret",
+            "4mRLtYMb6vq9qglomWEaT4ChxsXWcyqbQpuBNfMPOYOiDmYYQmjuaBsF2Zp0RyVeWkfqhE9TuGgAw7te",
```
This should be a const at the top of the file like `JUNO_PC_CLIENT_SECRET`

Gustash replied:
So should the CLIENT_ID imo

battledash replied:
Yeah, though that client ID might be used in multiple places, if it is put it in a common file next to `endpoints.rs`, so like `clients.rs` or something

battledash committed:
Token exchange for manual login

battledash committed:
Clean up a bit

battledash committed:
Slight fix

battledash committed:
Fix native and UI builds
